blog/2023-09-19-persisting-state/index.mdx

```diff
@@ -0,0 +1,136 @@
+image: /blog/2023-09-19-persisting-state.png
+---
+
+State machines have a useful feature: modeling state. But what happens when your user closes the browser and comes back later? Or when your server restarts? Or when your user logs out and logs back in? In this article, we'll learn how to persist and restore state in XState.
+
+Note: this article applies to XState v5 beta.
+
+## Quick reference
+
+If you're already familiar with XState, here's a quick reference for persisting and restoring state:
+
+```ts
+import { createActor } from 'xstate';
+import { someMachine } from './someMachine';
+
+// Get the state from localStorage (if it exists)
+const stateString = localStorage.getItem('some-state');
+
+// Create the state from the string (if it exists)
+const restoredState = stateString ? JSON.parse(stateString) : undefined;
+
+const actor = createActor(machine, {
+  // Restore the state (if it exists)
+  state: restoredState,
+});
+
+actor.subscribe(() => {
+  // Persist the state to localStorage
+  const persistedState = actor.getPersistedState();
+  localStorage.setItem('some-state', JSON.stringify(persistedState));
+});
+
+actor.start();
+```
+
+## Persisting and restoring state
+
+Let's say we have a state machine that represents a user's shopping cart:
+
+```ts
+import { createMachine, createActor, assign } from 'xstate';
+
+const checkoutMachine = createMachine({
+  id: 'checkout',
+  initial: 'shopping',
+  context: {
+    items: [],
+  },
+  states: {
+    shopping: {
+      on: {
+        'item.add': {
+          actions: assign({
+            items: ({ context, event }) => {
+              return [...context.items, event.item];
+            },
+          }),
+        },
+        checkout: 'checkingOut',
+      },
+    },
+    checkingOut: {
+      // ...
+    },
+  },
+});
+
+const checkoutActor = createActor(machine, {
+  // ...
+});
+```
+
+You may want to persist the state of this machine so that when the user comes back to the site, the items in their cart are still there, and the step in the checkout process is still the same. To do this, we need to remember to do two things:
+
+- **Persist the state** to some storage (e.g., localStorage, a database, etc.)
+- **Restore the state** from the storage when creating the actor
+
+First, you should determine your [persistence strategy](#persistence-strategies). For this example, we'll use [the `localStorage` API](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).
+
+Actors created with `createActor` have a `.getPersistedState()` method that returns the state that should be persisted. This state is a plain object that can be serialized to JSON using `JSON.stringify(persistedState)`. Note that this "persisted state" may be slightly different from the `snapshot` state returned from `actor.getSnapshot()`. Since the persisted state is a plain object, you can persist it to any storage that you want, as long as it can be restored as an object.
```

Comment:
```suggestion
Actors created with `createActor` have a `.getPersistedState()` method that returns the state that should be persisted. This state is a plain object that can be serialized to JSON using `JSON.stringify(persistedState)`. Note that this "persisted state" may be slightly different from the `snapshot` state returned from [`actor.getSnapshot()`](https://stately.ai/docs/xstate-v5/actors#actor-snapshots). Since the persisted state is a plain object, you can persist it to any storage that you want, as long as it can be restored as an object.
```